src/base/RefCount.h

```diff
@@ -27,10 +27,12 @@ class RefCount
 {
 
 public:
+    template<typename... Args>
+    inline static auto Make(Args&&... args) {
+        return RefCount<C>(new C(std::forward<Args>(args)...));
+    }
     RefCount () : p_ (nullptr) {}
 
-    RefCount (C const *p) : p_(p) { reference (*this); }
```

Comment:
IIRC we discussed this in the meeting last week. Prohibiting conversion from raw-pointer to Pointer for the `RefCountable` classes avoids all that extra annoyance with `new`.

FWIW, I intend to to make this constructor public again with a TODO for the transition once this PR passes review.

Reply:
> IIRC we discussed this in the meeting last week. Prohibiting conversion from raw-pointer to Pointer for the `RefCountable` classes avoids all that extra annoyance with `new`.

We did discuss the dangers of leaving `new` result in a raw pointer. I believe we agree that those dangers should be avoided. I hope we also agree that Make() is the right solution to avoid those dangers. However, all that does not imply that we should hide this conversion constructor.

```C++
C *rawP = new C; // a problem (should be prohibited but would be allowed, in some contexts, by this change)
Pointer p1 = new C; // not really a problem (but would be prohibited by this change)
Pointer p2 = rawP; // not really a problem (but would be prohibited by this change) or a minor out-of-scope problem
```
 
> FWIW, I intend to to make this constructor public again with a TODO for the transition once this PR passes review.

Noted, but we do not want to transition to a hidden conversion constructor, at least not as far as this PR, this problem, and Make() solutions are concerned. All those things require a very different transition goal and steps, the ones outlined in this change request.

Reply:
Sigh. I thought we had this settled.

The problem I am solving is that mixing explicit `rawP = new C;  ... ; Pointer p1 = rawP; ...` has problems for us determining whether `rawP` needs developers to code explicit `delete`(s) [in extra code paths handling `rawP` copies] or if the `Pointer` takes full control and does it implicitly in all code.

By explicitly prohibiting conversion between the types of pointer semantic, we clarify for reviewers and developers which memory management paradigm to apply to the variables (C++ RAII vs. C-style). Making the review process easier and requirements simpler to inform new developers about.


Essentially removing `new'/'delete' from the C++ language is an entirely different long-term situation which I am not going near tackling in this PR or any of its direct followups. I am leaving that to you after this transition converts most `RefCountable` child classes to `Make()` and highlights the ones with issues.

Reply:
> The problem I am solving is that mixing explicit `rawP = new C; ... ; Pointer p1 = rawP; ...` has problems for us determining whether `rawP` needs developers to code explicit `delete`(s) [in extra code paths handling `rawP` copies] or if the `Pointer` takes full control and does it implicitly in all code.

The `rawP = new C; ... ; Pointer p1 = rawP` problem is solved by migrating to Make(). After that migration, `rawP = new C` disappears completely. Hiding this conversion constructor does not help with that migration and does not enforce that migration, but creates other problems. Some of those other problems are rather complex/controversial. _This_ PR does not need to open that Pandora box.

Migration from `const auto rawP = new C()` to the equivalent of `const auto p = Make()` does not solve all pointer-related problems, of course. If you would like to solve some other specific problem, let's discuss, but that discussion does not need to block this PR AFAICT _and_ should be based on Make()-using code (as if this PR has been merged).


> By explicitly prohibiting conversion between the types of pointer semantic, we clarify for reviewers and developers which memory management paradigm to apply to the variables (C++ RAII vs. C-style). Making the review process easier and requirements simpler to inform new developers about.

Agreed. Implicit conversions do have significant costs. On the other hand, we should not prohibit all useful (and occasionally essential) code that uses raw pointers correctly. Raw pointers are the best solution for some problems. Again, I am happy to discus solutions for reducing incorrect use of raw pointers in post-Make() code. However, I insist on formulation of those problems (if they are to be discussed) using post-Make() code templates/examples. The current example quoted above is pre-Make(). Such examples will not exist (i.e. become invalid) after we convert to Make().

AFAICT, there is basic consensus regarding conversion to Make(), so we should assume or capitalize on that great simplification/improvement when discussing further pointer-related improvements.